### ISSUE TITLE
[Proposal] Update Namespaces to Match Folder Structure

### ISSUE THREAD

**Issue by brminnick** [CommunityToolkit/Maui#219]
# Update Namespaces to Match Folder Structure

* [x] Proposed
* [x] Prototype
* [ ] Implementation: Not Started
  * [ ] iOS Support
  * [ ] Android Support
  * [ ] macOS Support
  * [ ] Windows Support
* [ ] Unit Tests: Not Started
* [ ] Sample: Not Started
* [ ] Documentation: Not Started

## Link to Discussion

Discussion on Namespaces Matching Folder Structure
- https://github.com/CommunityToolkit/Maui/pull/214/files#r768099127
Discussion on Moving Enums to a folder called `Enums`, and updating their namespaces accordingly
- https://github.com/CommunityToolkit/Maui/pull/214/files#r767382464

## Summary
[summary]: #summary

This Proposal solves two problems:
- Ensures a file's namespace matches its folder hierarchy 
- Moves Enums to their own File in an `Enums` folder

Because this is a breaking change that requires XAML users to add additional `xmlns`, this Proposal will be placed on hold until a Toolkit-wide namespace is introduced in  #144, e.g. `xmlns:toolkit="http://schemas.microsoft.com/dotnet/2021/maui/tookit"`.

## Motivation
[motivation]: #motivation

Currently, both `CommunityToolkit.Maui` and `CommunityToolkit.Maui.Core` use the root folder for their namespace, but nested folders are not included. 

E.g. `/Behaviors/EventToCommandBehavior.shared.cs` uses `namespace CommunityToolkit.Maui.Behaviors;`
E.g. `/Behaviors/Validators/CharactersValidationBehavior.shared.cs` also uses `namespace CommunityToolkit.Maui.Behaviors;`

This root-folder namespace matches the precedent set in [`Xamarin.CommunityToolkit`](https://github.com/xamarin/xamarincommunitytoolkit/), however, we would like to follow the industry-standard of matching the folder hierarchy.

## Detailed Design
[design]: #detailed-design

- Update Namespaces to match folder structure
  - E.g. `/Behaviors/Validators/CharactersValidationBehavior.shared.cs` will become `namespace CommunityToolkit.Maui.Behaviors.Validators;`
- Add [`IDE0130`](https://hanskindberg.wordpress.com/2021/08/27/net-analyzers-ide0130/) to `DirectoryBuild.props`
  - `<WarningsAsErrors>nullable,CS1574,IDE0130</WarningsAsErrors>`
  - This ensures that all future PRs will have namespaces that match the file's folder hierarchy
-  Create `Enums` Folder, and move each `enum` types into their own file in this folder
  - Currently, all single-use Enums are placed in the same file as the class that uses them

## Drawbacks
[drawbacks]: #drawbacks

- Breaks precedence set by Xamarin.CommunityToolkit
- Will require XAML users to set an additional `xmlns`
  - This is averted by holding this Proposal until #144 is completed
- Will require C# users to declare an additional `using CommunityToolkit.Maui.*` namespace

## Alternatives
[alternatives]: #alternatives

There are two other alternatives:
1. Matching the Assembly namespace
  - Every file in `CommunityToolkit.Maui` would use `namespace CommunityToolkit.Maui`
  - Every file in `CommunityToolkit.Maui.Core` would use `namespace CommunityToolkit.Maui.Core`
2. Matching `Xamarin.CommunityToolkit`'s Precedence of Root-Folder Namespaces

## Unresolved Questions
[unresolved]: #unresolved-questions

None that I can think of

**Comment by bijington** [comment]
@brminnick I see you have championed this but I am happy to provide the effort to get this done if needed

**Comment by pictos** [comment]
I would like to discuss this proposal a little bit more. 
A bunch of .NET libs follows the pattern that we're using and they work very well. I can list some pros in keeping our current structure, that's the namespace doesn't follow the folder structure.

- APIs, will be easy to discover by users;
- We can save some `using` imports;
- Lees `xmlns` declaration in XAML;
- This is very common in the Xamarin world.

With that can't see a good reason to perform this change.

**Comment by brminnick** [comment]
> With that can't see a good reason to perform this change

I don't disagree. My personal preference is to use assembly name for each namespace. E.g. `/Behaviors/Validators/CharactersValidationBehavior.shared.cs` would use `namespace CommunityToolkit.Maui`.

I've always found it a bit clunky to use a different namespace for each folder.

**Comment by pictos** [comment]
> > With that can't see a good reason to perform this change
> 
> I don't disagree. My personal preference is to use assembly name for each namespace. E.g. `/Behaviors/Validators/CharactersValidationBehavior.shared.cs` would use `namespace CommunityToolkit.Maui`.
> 
> I've always found it a bit clunky to use a different namespace for each folder.

I can add the Xamarin.Essentials, for example, there's one namespace for APIs and that's very easy to discover what APIs you've on the lib.

**Comment by bijington** [comment]
We certainly don't have to go ahead with this proposal, I guess it comes naturally to me to have folders/namespaces match but that is probably just because I have worked that way for so long 😄 

I do like some level of categorisation which I believe we would get with the old XCT approach where it isn't too easy to discover everything and it resulting in more noise than being helpful. I think we are also moving some bits around like in #35 where animation related bits didn't fit in behaviors. 

Basically I am really good at sitting on the fence 😄

**Comment by davidbuckleyni** [comment]
This seems a simple enough change. I think i bit more of can chew with main libarys is this one up for grabs.

**Comment by pictos** [comment]
@davidbuckleyni not yet, we're still discussing how we would like to implement this

**Comment by brminnick** [comment]
[Reassigning Champion to @bijington](https://github.com/CommunityToolkit/Maui/issues/219#issuecomment-992916268)

**Comment by bijington** [comment]
How do we feel about proceeding with this change? I see we have mixed opinions

**Comment by brminnick** [comment]
Good question! 

Looks like there are 3 Options being discussed:
1. Matching the Folder Structure
2. Matching only the Assembly namespace
    - Every file in `CommunityToolkit.Maui` would use `namespace CommunityToolkit.Maui`
    - Every file in `CommunityToolkit.Maui.Core` would use `namespace CommunityToolkit.Maui.Core`
3. Matching the Root-Folder Namespace
    - This matches `Xamarin.CommunityToolkit`'s Precedence of Root-Folder Namespaces

Let's do a vote:

1. React with a ❤️ to vote for Option 1
    - Matching the Assembly + Folder Structure
2. React with a 🚀 to vote for Option 2
    - Matching only the Assembly namespace
3. React with a 👀 to vote for Option 3
    - Matching the Root-Folder Namespace

**Comment by bijington** [comment]
@brminnick just to confirm for `CommunityToolkit.Maui.Core/Views/Popup/Popup.macios.cs` (https://github.com/CommunityToolkit/Maui/blob/main/src/CommunityToolkit.Maui.Core/Views/Popup/Popup.macios.cs) 

| Option 1 | Option 2 | Option 3 |
| --- | --- | --- |
| `CommunityToolkit.Maui.Core.Views.Popup.Popup` | `CommunityToolkit.Maui.Core.Popup` | `CommunityToolkit.Maui.Core.Popup` |

I am not sure I understand the difference between 2 and 3. Perhaps this is a bad example?

**Comment by VladislavAntonyuk** [comment]
If we go with option 2 or 3, please fix editorconfig to remove warning

**Comment by VladislavAntonyuk** [comment]
@bijington 
| Option 1 | Option 2 | Option 3 |

| --- | --- | --- |

| `CommunityToolkit.Maui.Core.Views.Popup` | `CommunityToolkit.Maui.Core` | `CommunityToolkit.Maui.Core.Views` |

**Comment by brminnick** [comment]
FYI - I updated my comment above to include an example 👍

**Comment by michael-hawker** [comment]
In the Windows Community Toolkit we've found that flattening namespaces down helps with reducing the amount of usings required in discovering what's available to use. This is also important for anything that needs to be included in XAML as having a bunch of xmlns is confusing and unwieldly, so we've been trying to reduce that as well.

**Comment by pictos** [comment]
> In the Windows Community Toolkit we've found that flattening namespaces down helps with reducing the amount of usings required in discovering what's available to use. This is also important for anything that needs to be included in XAML as having a bunch of xmlns is confusing and unwieldly, so we've been trying to reduce that as well.

@michael-hawker for XAML we have been using xmlns schema, as you can see [here](https://github.com/xamarin/XamarinCommunityToolkit/blob/main/src/CommunityToolkit/Xamarin.CommunityToolkit/AssemblyInfo/AssemblyInfo.shared.cs#L7-L11) not use if it's possible on Windows world

**Comment by bijington** [comment]
> > In the Windows Community Toolkit we've found that flattening namespaces down helps with reducing the amount of usings required in discovering what's available to use. This is also important for anything that needs to be included in XAML as having a bunch of xmlns is confusing and unwieldly, so we've been trying to reduce that as well.
> 
> @michael-hawker for XAML we have been using xmlns schema, as you can see [here](https://github.com/xamarin/XamarinCommunityToolkit/blob/main/src/CommunityToolkit/Xamarin.CommunityToolkit/AssemblyInfo/AssemblyInfo.shared.cs#L7-L11) not use if it's possible on Windows world

You beat me to it :). As per this discussion #144

**Comment by michael-hawker** [comment]
Yeah, definitely missing that attribute in the UWP/WinUI 3 space; that'd have made migration a lot easier between the two... 🙁

**Comment by sander1095** [comment]
I feel like the examples should be updated to show 3 different usings from different folders. That way example 1 has 3 usings, while example 2 has 1

**Comment by vhugogarcia** [comment]
My personal recommendation will be:

MauiCommunity.* 

And match folder structure. I believe the word toolkit stands for UI elements, however, I believe the community is strong now and MauiCommunity makes more sense to me. By setting MauiCommunity.* we recognize people around the world supporting this amazing library. 

We are all **MauiCommunity** 🤝😎

**Comment by danardelean** [comment]
My 2 cent: from a "commercial" point of view and sometimes for the people that make the decision if the team can use or not the Community Toolkit I think it would be better to drop the Community term from the namespace. 

While we all know that it is built by the community (and the quality of the code that gets merged is the same as the framework) leaving only Toolkit would give it more "commercial" trust in my opinion.

It would be :
namespace Maui.Toolkit.Behaviors.Validators;

**Comment by bijington** [comment]
I'm just realising that if we go ahead with #144 then only option 2 would give the same consistent intellisense results across C# and XAML.

**Comment by Sergio0694** [comment]
> "any insights from the .NET Community Toolkit you'd want to share?"

Based on what we've been doing iin both the Windows Community Toolkit and the .NET Community Toolkit, we essentially used the assembly name as the root namespace in the vast majority of cases, with additional nested namespaces in few cases where it made sense to further categorize public APIs. FWIW, I personally think that setting a rule that the namespace should be coupled with the folder structure (which seems to be the most upvoted poll option right now) is wrong. That means the internal organization of the code is tied to the public API surface you expose, which should not be the case. For instance, putting extensions in an "Extension" namespace is wrong, but with this approach you're then forced to place all extensions in the parent directory, making the project tree more cluttered, instead of just having a nice "Extensions" folder to keep things organized internally.

To make another example, almost all our APIs in the `Microsoft.Toolkit.Uwp.UI.Animations` package are in the `Microsoft.Toolkit.Uwp.UI.Animations` namespace. If we had matched the namespace with the folder structure, this means we would've had to just put **all** files into the same root folder, which would've been much harder to navigate as a resul. For instance, Roslyn does this too, where they just organize code internally without matching the namespace.

There is also an arguing for preserving some granularity with sub-namespaces, in case you wanted to avoid making source breaking changes if you decided to split parts of the package into separate assemblies in the future (eg. keeping all the behaviors into a ".Behaviors" nested namespace). This might or might not be a concern.

**Comment by bijington** [comment]
@Sergio0694 thanks for sharing that. I agree with your points which leads me on to thinking perhaps an option 4 is best being a hybrid of 2 and 3. Where we keep the namespaces as short as possible but at times we feel categorising is valuable do it.

I suspect option 3 may cover it in the majority if not all existing samples anyway.
